pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,481 @@
+		ip, cidr, err := net.ParseCIDR(mcastGroupAddress)
+		if err != nil {
+			ip = net.ParseIP(mcastGroupAddress)
+			cidr = &net.IPNet{}
+			cidr.IP = ip
+			cidr.Mask = net.CIDRMask(32, 32)
+		}
+		mcastGroupAddressSet.Insert(mcastGroupAddress)
+		if !exists {
+			item := mcastItem{
+				groupAddress: *cidr,
+				ruleIDs:      make(map[string]*uint16),
+			}
+			item.ruleIDs[ruleID] = priority
+			c.mcastItemRuleIDMap[mcastGroupAddress] = item
+		} else {
+			item.ruleIDs[ruleID] = priority
+			c.mcastItemRuleIDMap[mcastGroupAddress] = item
+		}
+	}
+	if mcastGroupAddressSet.Len() > 0 {
+		fmt.Println(mcastGroupAddressSet)
+		c.ruleIDGroupAddressMap[ruleID] = mcastGroupAddressSet
+	}
+	g := mcastGroupEvent{
+		eType: rulechanged,
+	}
+	c.eventCh <- g
+}
+
+func (c *multicastController) updateGroupAddressForTableIDs(ruleID string, priority *uint16, mcastGroupAddresses []string) {
+	c.mcastItemMutex.Lock()
+	defer c.mcastItemMutex.Unlock()
+
+	c.ruleIDGroupMapMutex.Lock()
+	defer c.ruleIDGroupMapMutex.Unlock()
+
+	staleMcastGroupAddresses, ok := c.ruleIDGroupAddressMap[ruleID]
+	for _, mcastGroupAddress := range mcastGroupAddresses {
+		item, exists := c.mcastItemRuleIDMap[mcastGroupAddress]
+		ip, cidr, err := net.ParseCIDR(mcastGroupAddress)
+		if err != nil {
+			ip = net.ParseIP(mcastGroupAddress)
+			cidr = &net.IPNet{}
+			cidr.IP = ip
+			cidr.Mask = net.CIDRMask(32, 32)
+		}
+		if !exists {
+			item := mcastItem{
+				groupAddress: *cidr,
+				ruleIDs:      make(map[string]*uint16),
+			}
+			item.ruleIDs[ruleID] = priority
+			c.mcastItemRuleIDMap[mcastGroupAddress] = item
+		} else {
+			item.ruleIDs[ruleID] = priority
+			c.mcastItemRuleIDMap[mcastGroupAddress] = item
+		}
+	}
+	if !ok {
+		staleMcastGroupAddresses = sets.String{}
+	}
+	for staleGroupAddress := range staleMcastGroupAddresses {
+		if item, ok := c.mcastItemRuleIDMap[staleGroupAddress]; ok {
+			if _, ok = item.ruleIDs[ruleID]; ok {
+				delete(item.ruleIDs, ruleID)
+				if len(item.ruleIDs) > 0 {
+					c.mcastItemRuleIDMap[staleGroupAddress] = item
+				} else {
+					c.cleanupGroupAddressForTableIDsUnlocked(staleGroupAddress)
+				}
+			}
+		}
+	}
+
+	newMcastGroupAddresses := sets.String{}
+	for _, mcastGroupAddress := range mcastGroupAddresses {
+		newMcastGroupAddresses.Insert(mcastGroupAddress)
+	}
+	c.ruleIDGroupAddressMap[ruleID] = newMcastGroupAddresses
+	g := mcastGroupEvent{
+		eType: rulechanged,
+	}
+	c.eventCh <- g
+}
+
+func (c *multicastController) deleteGroupAddressForTableIDs(ruleID string, groupAddresses []string) {
+	c.mcastItemMutex.Lock()
+	defer c.mcastItemMutex.Unlock()
+	c.ruleIDGroupMapMutex.Lock()
+	defer c.ruleIDGroupMapMutex.Unlock()
+	for _, groupAddress := range groupAddresses {
+		item, exists := c.mcastItemRuleIDMap[groupAddress]
+		klog.V(2).Infof("deleteGroupAddressForTableIDs groupAddress: %v, exist %v, map %+v %+v",
+			groupAddress, exists, c.mcastItemRuleIDMap, item)
+		if _, ok := item.ruleIDs[ruleID]; exists && ok {
```

Comment:
item is nil if exists==false, then this line should cause a nil pointer crash?

Reply:
done